protos/encodings21.proto

```diff
@@ -0,0 +1,444 @@
+  // there is never a reason to use a bits per value that is not a multiple of 8.  The complexity is not
+  // worth the small savings in space since this encoding is typically used with large values already.
+  oneof details {
+    // If this is a fixed width block then we need to have a fixed number of bits per value
+    uint32 bits_per_value = 3;
+    // If this is a variable width block then we need to have a fixed number of bits per offset
+    uint32 bits_per_offset = 4;
+  }
+  // The number of items in the page
+  uint32 num_items = 5;
+  // The number of visible items in the page
+  uint32 num_visible_items = 6;
+  // Description of the compression of values
+  CompressiveEncoding value_compression = 7;
+  // The meaning of each repdef layer, used to interpret repdef buffers correctly
+  repeated RepDefLayer layers = 8;
+}
+
+// A layout used for pages where all values are null
+//
+// There may be buffers of repetition and definition information
+// if required in order to interpret what kind of nulls are present
+message AllNullLayout {
+  // The meaning of each repdef layer, used to interpret repdef buffers correctly
+  repeated RepDefLayer layers = 5;
+}
+
+// Describes the structural encoding of a page
+message PageLayout {
+  oneof layout {
+    // A layout used for pages where the data is small
+    MiniBlockLayout mini_block_layout = 1;
+    // A layout used for pages where all values are null
+    AllNullLayout all_null_layout = 2;
+    // A layout used for pages where the data is large
+    FullZipLayout full_zip_layout = 3;
+  }
+}
+
+// # Compressive Encodings
+//
+// These encodings describe how an array is compressed.  An encoding may split an
+// array into multiple buffers.  The buffers can then be compressed further (and split
+// into yet more buffers).  The entire process forms a tree of encodings with the root
+// of the tree being the initial array and the leaves being the final compressed buffers.
+//
+// # Data blocks and buffers
+//
+// Data blocks are a simplified version of arrays and represent a collection of buffers grouped
+// with some kind of interpretation.  Data blocks are the input and output of compressive encodings.
+// There are different kinds of data blocks:
+// - Fixed width data blocks (e.g. u8, u16, ...)
+// - Variable width data blocks (e.g. strings, binary)
+// - Struct data blocks (note: this is for packed structs, normal structs are encoded in the structural encoding)
+//
+// In addition, leaf encodings may output "buffers".  These are fully compressed buffers of data that
+// are stored in the page and no longer compressed.
+
+// Compression applied to a single buffer of data
+//
+// A buffer is the leaf of the compression tree.  Unlike data blocks, which can
+// be further compressed with a variety of techniques, a buffer cannot be understood
+// in any particular way.
+//
+// A general compression scheme may be applied to a buffer.  This is something like
```

Comment:
Why this is not something like a enum?

Reply:
Ok, I switched to enum in https://github.com/lancedb/lance/pull/4434/commits/e134f6c602a14276db7220f5d8e6ee8114d5eab8